### ISSUE TITLE
Timeout rejections are unhandled

### ISSUE THREAD

**Issue by moltar** [amplitude/experiment-js-client#134]
In this code:

https://github.com/amplitude/experiment-js-client/blob/05c2c3898cd7dcac4b60ae4794dd73c7e7004be3/packages/experiment-browser/src/transport/http.ts#L18-L36

The timeout rejections will be unhandled, polluting the console with errors, and error reporting tools such as Sentry get flooded with useless errors.

The [comment on GH](https://github.com/github/fetch/issues/175#issuecomment-284787564) from which this function was copied from states for this to be safe, but indeed it is not.

A small repro is this [codesandbox](https://codesandbox.io/p/sandbox/2r9zz7?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm2ugug1t00062e69uhb5ipju%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm2ugug1t00022e693hnextop%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm2ugug1t00032e69s5g1wzza%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm2ugug1t00052e69g4topvzn%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm2ugug1t00022e693hnextop%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm2ugug1t00012e69losme2mw%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%257D%255D%252C%2522id%2522%253A%2522cm2ugug1t00022e693hnextop%2522%252C%2522activeTabId%2522%253A%2522cm2ugug1t00012e69losme2mw%2522%257D%252C%2522cm2ugug1t00052e69g4topvzn%2522%253A%257B%2522id%2522%253A%2522cm2ugug1t00052e69g4topvzn%2522%252C%2522activeTabId%2522%253A%2522cm2ugutgw00172e69dog5uh1e%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm2ugug1t00042e69xf0az4im%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm2ugutgw00172e69dog5uh1e%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm2ugug1t00032e69s5g1wzza%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cm2ugug1t00032e69s5g1wzza%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Afalse%252C%2522sidebarPanelSize%2522%253A15%257D).

A less naive implementation can be found in this package: https://github.com/sindresorhus/p-timeout

**Comment by tyiuhc** [comment]
@moltar, many thanks for raising this issue. To reduce error-log pollution, the specific timeout errors are now logged at the `debug`-level in v1.12.0. Please let us know if this issue persists!

**Comment by moltar** [comment]
I don't think it addresses the main problem - which is unhandled exceptions due to naive handling of the promises.

Please refer to my repro example.

**Comment by bgiori** [comment]
Hi @moltar, thanks again for submitting this issue.

Question regarding the code sandbox example you provided. 

```
try {
  timeout(sleep(2_000), 1_000);
} catch {
  console.log("timeout");
}
```

Is the expected behavior that the catch statement is run? 

If so, this is expected to not occur since the timeout function returns a promise. To catch the error from the promise, we either need to `await` the promise returned by `timeout` with the existing try/catch, or use a `.catch()` function chained to `timeout(...)`.

The caller of `request` which calls `timeout` is set up to handle the errors from the rejected promises. Previously we were logging all these errors including the timeouts as `error` logs, but now we're logging them as `debug`.

Does this resolve the issue? Or did I miss something with respect to how the timeout function returns it's errors?

**Comment by tyiuhc** [comment]
@moltar thank you for your patience regarding this issue. It is possible that the continued Sentry errors are due to timeout errors not being caught for flag fetches when the client is not in `debug`-logging mode. v1.12.3 will now always catch timeout errors, regardless of logging mode. Please let us know if your issue persists.
